session/session.go

```
@@ -90,6 +108,152 @@ func Open(config sfdc.Configuration) (*Session, error) {
 	return session, nil
 }
 
+func openDeviceSession(config sfdc.Configuration) (*Session, error) {
+	request, err := buildDeviceAuthenticationFlowInitiationRequest(config.Credentials)
+	if err != nil {
+		return nil, fmt.Errorf("failed to build ")
+	}
+
+	initResp, err := makeDeviceAuthenticationFlowInitiationRequest(request, config.Client)
+	if err != nil {
+		return nil, err
+	}
+
+	var tokenResp *accessTokenResponse
+	var done = make(chan struct{})
+
+	go func() {
+		for {
+			select {
+			case <-time.After(time.Duration(initResp.Interval+1) * time.Second):
+				accessTokenReq, err := buildDeviceAuthenticationFlowAccessRequest(config.Credentials, initResp)
+				if err != nil {
+					log.Fatalf("failed to build device authentication flow access request: %s", err)
+				}
+
+				// Make authorization request
+				var tokenErrResp *deviceAccessTokenErrorResponse
+				if tokenResp, tokenErrResp, err = makeDeviceAuthenticationFlowAccessTokenRequest(accessTokenReq, config.Client); err != nil {
+					log.Fatalf("error polling for device access token: %s", err)
+				}
+
+				// Success!
+				if tokenResp != nil {
+					close(done)
+					return
+				}
+
+				if tokenErrResp != nil && tokenErrResp.Error == AuthorizationPendingErrorCode {
+					log.Printf("Authorization pending. Please enter code '%s' at %s to authorize application.", initResp.UserCode, initResp.VerificationURI)
```

we should return these errors back